spec/validator_spec.js

```diff
@@ -20,183 +20,212 @@ function validate(xmlData, error, line = 1) {
-        validate("<rootNode#@aa></rootNode#@aa>", {
+        validateIgnoringNS("<rootNode#@aa></rootNode#@aa>", {
             InvalidTag: "Tag 'rootNode#@aa' is an invalid name."
         });
     });
 
     it("should return false for non xml text", function () {
-        validate("rootNode", {
+        validateIgnoringNS("rootNode", {
             InvalidChar: "char 'r' is not expected."
         });
     });
 
     it("should validate self closing tags", function () {
-        validate("<rootNode><validtag1  /><validtag2/><validtag3  with='attrib'/><validtag4 />text<validtag5/>text</rootNode>");
+        validateIgnoringNS("<rootNode><validtag1  /><validtag2/><validtag3  with='attrib'/><validtag4 />text<validtag5/>text</rootNode>");
     });
 
     it("should not consider these as self closing tags", function () {
-        validate("<rootNode><validtag1/><invalid tag/><validtag3  with='attrib'/></rootNode>", {
+        validateIgnoringNS("<rootNode><validtag1/><invalid tag/><validtag3  with='attrib'/></rootNode>", {
             InvalidAttr: "boolean attribute 'tag' is not allowed."
         });
-        validate("<rootNode><notSelfClosing/ ></rootNode>", {
+        validateIgnoringNS("<rootNode><notSelfClosing/ ></rootNode>", {
             InvalidAttr: "Attribute '/' has no space in starting."
         });
     });
 
 
     it("should not validate xml string when closing tag is different", function () {
-        validate("<rootNode></rootnode>", {
+        validateIgnoringNS("<rootNode></rootnode>", {
             InvalidTag: "Closing tag 'rootNode' is expected inplace of 'rootnode'."
         });
     });
 
     it("should not validate xml string when closing tag is invalid", function () {
-        validate("<rootNode>< /rootnode>", {
+        validateIgnoringNS("<rootNode>< /rootnode>", {
             InvalidTag: "There is an unnecessary space between tag name and backward slash '</ ..'."
         });
-        validate("<rootNode></ rootnode>", {
+        validateIgnoringNS("<rootNode></ rootnode>", {
             InvalidTag: "There is an unnecessary space between tag name and backward slash '</ ..'."
         });
-        validate("<rootNode></rootnode 123>", {
+        validateIgnoringNS("<rootNode></rootnode 123>", {
             InvalidTag: "Closing tag 'rootnode' can't have attributes or invalid starting."
         });
     });
 
     it("should validate simple xml string with namespace", function () {
-        validate("<root:Node></root:Node>");
+        validateWithNS("<root:Node xmlns:root='urn:none'></root:Node>");
+    });
+
+    it("should not validate simple xml string when namespace is not defined", function () {
+        validateWithNS("<root:Node></root:Node>", {
+            InvalidNS: "Namespace prefix 'root' is not defined for tag 'root:Node'"
+        });
+    });
+
+    it("should not validate xml when namespace is defined later", function () {
```

Comment:
Probably we need more tests to check 
1. self-closing tags
2. sibling tags
3. multiple NS for a tag
4. NS in attribute name

Reply:
Added tests. However, the behavior for point 3 might be changed once [discussion](https://github.com/NaturalIntelligence/fast-xml-parser/pull/275#discussion_r449852951) on multiple colons is closed.

Reply:
What I mean with multiple NS for a tag means `<root xmlns:ab="url" xmlns:bc="" />`

We'll also have to check the duplicate attributes. `<root xmlns:ab="url" xmlns:bc="" ab:att="" bc:att="" />` when `ignoreAttributes` is true or false.

Reply:
For multiple attributes with same name, expecting the following behavior:
1. If namespace is not ignored, multiple attributes with same name and same namespace prefix are considered duplicate but attributes with same name and different namespace prefix are considered as different.
2. If namespace is ignored, multiple attributes with same name are considered duplicate (irrespective of namespace prefix).
Hope this looks fine.
Also, ignoreAttributes is currently not used for validation. Should we introduce this as well?

Reply:
Yes for both

Reply:
To be honest, I have actually implemented all those things in a different implementation of FXP 2 months ago. I've re-written the parser. But I'm not getting sufficient time to complete the UTs. So not publishing them.

Reply:
Pushed the changes for now. Hope you get the time soon. Would love to see the new implementation